File: GofExperience/src/Composite/Composite.cpp
#include "../stdafx.h"
#include "Composite.h"
#include "Leaf.h"

void SafeDelete(Component *component)
{
	if (component != NULL)
	{
		delete component;
	}
}

Composite::Composite(const char *name)
	: Component(name)
{
}

Composite::~Composite()
{
	Clear();
}

string Composite::GetPath(void) const
{
	return Component::GetPath() + "/";
}

void Composite::Add(Component *component)
{
	components.push_back(component);
	component->SetParent(this);
}

void Composite::Remove(Component *component)
{
	if (component != NULL)
	{
		components.remove(component);
		component->SetParent(NULL);
		delete component;
	}
}

void Composite::Clear(void)
{
	for_each(components.begin(), components.end(), SafeDelete);
}

void CompositeTest(void)
{
	Composite *root = new Composite("usr");
	Composite *bin = new Composite("bin");
	Leaf *perl = new Leaf("perl");
	root->Add(bin);
	bin->Add(perl);
	cout << perl->GetPath().c_str() << endl;
}

File: GoFExperienceCS/Structural/Composite/Composite.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace GoFExperienceCS.Structural.Composite
{
	sealed class Composite
		: Component
	{

		private readonly ICollection<Component> childs = new List<Component>();

		private readonly string[] dump;

		public Composite(string name)
			: base(name)
		{
			dump = new string[] { Name + "/" };
		}

		public T Add<T>(T component)
			where T : Component
		{
			if (component != null)
			{
				childs.Add(component);
			}
			return component;
		}

		public override IEnumerable<string> GetDump()
		{
			var path = dump[0];
			var result = childs.SelectMany(c => c.GetDump(), (co, l) => path + l);
			return dump.Concat(result);
		}

		private static void Main()
		{
			var root = new Composite("");
			root.Add(new Leaf("boot"));
			root.Add(new Composite("bin")).Add(new Leaf("ls"));
			var usr = root.Add(new Composite("usr"));
			usr.Add(new Composite("local"));
			usr.Add(new Composite("bin")).Add(new Leaf("perl"));
			usr.Add(new Composite("share")).Add(new Leaf("redmine"));
			usr.Add(new Composite("src"));
			root.Add(new Leaf("sbin"));
			root.Add(new Leaf("swap"));
			foreach (var p in root.GetDump())
			{
				Console.WriteLine(p);
			}
			Console.ReadKey();
		}
	}
}
